MB-4830 Bug initial authorized weight should be same as weight allowance

donaldthai opened the pull request:
## Description

> According to LTC Ryan, weight entitlement/allowance should not include pro gear. We still want to display the amounts for pro gear and spouse pro gear, like we already to in the TOO interface. However, only rank and dependents should be used to determine the initial weight allowance and authorized allowance. 
> 
> ## Acceptance Criteria
> 
> Write a test to verify the existing broken behavior
> 
> Make the test pass by modifying the `GetEntitlement` function to only take the rank and dependents and return the appropriate number
> 
> Modify all the places where we call `GetEntitlement`. 

## Reviewer Notes

Is there anything you would like reviewers to give additional scrutiny?

## Setup

Add any steps or code to run in this section to help others prepare to run your code:

```sh
make db_dev_e2e_populate
make server_run
make office_client_run
```

## To test
1. Log in to office app as TOO role, locally
2. Click on any test move (non of the moves should have a set authorized weight)
3. Scroll down to `Allowances`
![image](https://user-images.githubusercontent.com/1522549/100281512-e906e680-2f1e-11eb-879d-4ce9c102ece0.png)
4. Weight Allowance and Authorized Weight should be the same

Notes:
- To test for dependents_authorized is false, you can quickly change the value in the database to see the changes in the `entitlements` table

## Code Review Verification Steps


* [ ] Code follows the guidelines for [Logging](https://github.com/transcom/mymove/wiki/Backend-Programming-Guide#logging)
* [ ] The requirements listed in
 [Querying the Database Safely](https://github.com/transcom/mymove/wiki/Backend-Programming-Guide#querying-the-database-safely)
 have been satisfied.
* [ ] Request review from a member of a different team.
* [ ] Have the Jira acceptance criteria been met for this change?

## References

* [Jira story](https://dp3.atlassian.net/browse/MB-4830) for this change

donaldthai commented on pkg/handlers/internalapi/moves_test.go:
```diff
@@ -354,7 +354,6 @@ func (suite *HandlerSuite) TestShowMoveDatesSummaryHandler() {
 	suite.Equal(moveDate, *okResponse.Payload.MoveDate)
 
 	pack := []strfmt.Date{
-		strfmt.Date(time.Date(2018, 10, 4, 0, 0, 0, 0, time.UTC)),
```
Was able to remove this because the weights are now based on `TotalWeightSelf` or `TotalWeightSelfWithDependents` which is a lot less than with pro-gear and pro-gear spouse. The formula used is an estimated `5000lbs` per day. So a rank E1 with dependents entitlement weight is 8000, that would mean an estimated 2 days for packing.

robot-mymove commented:
<!--
  0 failure: 
  0 warning: 
  1 messages
  
  DangerID: danger-id-Danger;
-->



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>:link: <a href="https://dp3.atlassian.net/browse/MB-4830">MB-4830</a></td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 67a4deb36d6569b42686155f8c71a24a3e79e2d8
</p>

noelledusahel commented on pkg/handlers/internalapi/entitlements_test.go:
```diff
@@ -40,8 +40,8 @@ func (suite *HandlerSuite) TestValidateEntitlementHandlerReturns200() {
 	move := ppm.Move
 
 	// When: rank is E1, the orders have dependents and spouse gear, and
```
nit: is this comment still accurate re: including pro gear?

donaldthai replied:
Nice catch! I'll remove it.

monfresh commented:
@donaldthai I think there's still some missing logic. For an E-1 rank with dependents (which I think is what most moves locally have), both weight allowance and authorized weight should be 8000. I fixed the weight allowance in my branch, so if you could please merge master into your branch, we can verify that here as well.

Before, if the authorized weight was empty in the DB, like it is in all the devseed moves, then it would fall back to `TotalWeightSelf`, which is 5000, which is wrong when there are dependents. There should be the same conditional I added in my PR.

donaldthai commented:
> @donaldthai I think there's still some missing logic. For an E-1 rank with dependents (which I think is what most moves locally have), both weight allowance and authorized weight should be 8000. I fixed the weight allowance in my branch, so if you could please merge master into your branch, we can verify that here as well.
> 
> Before, if the authorized weight was empty in the DB, like it is in all the devseed moves, then it would fall back to `TotalWeightSelf`, which is 5000, which is wrong when there are dependents. There should be the same conditional I added in my PR.

@monfresh what a coincidence! I just pushed the latest code fixing the exact thing you're talking about!

monfresh commented on pkg/models/ghc_entitlements.go:
```diff
@@ -42,6 +42,9 @@ func (e *Entitlement) AuthorizedWeight() *int {
 	case e.WeightAllotment() != nil:
+		if e.DependentsAuthorized != nil && *e.DependentsAuthorized == true {
+			return &e.WeightAllotment().TotalWeightSelfPlusDependents
+		}
```
Can we update the comment too, please? It still refers to `TotalWeightSelf` only

donaldthai replied:
Done!

monfresh commented on pkg/models/ghc_entitlements_test.go:
```diff
@@ -1,25 +1,37 @@
 
-	suite.Equal(entitlement.WeightAllotment().TotalWeightSelf, *entitlement.AuthorizedWeight())
+		suite.Equal(entitlement.WeightAllotment().TotalWeightSelfPlusDependents, *entitlement.AuthorizedWeight())
+	})
```
Can we use plain models here without saving them to the DB since we're only testing logic based on attributes?

donaldthai replied:
Oh yea, that's definitely a good route to go with. I can refactor this.

donaldthai replied:
Yup, looks like removing the db create call does the trick. I'll push up the code.

monfresh commented on pkg/notifications/move_submitted.go:
```diff
@@ -69,7 +69,7 @@ func (m MoveSubmitted) emails(ctx context.Context) ([]emailContent, error) {
 	}
 
-	totalEntitlement, err := models.GetEntitlement(*serviceMember.Rank, orders.HasDependents, orders.SpouseHasProGear)
+	totalEntitlement, err := models.GetEntitlement(*serviceMember.Rank, orders.HasDependents)
```
In what context does this get used? I wonder if this is expected to include pro gear like in the summary worksheet?

donaldthai replied:
I think this refers to when a move is submitted and an email is sent to the service member. @kimallen knows more though.

donaldthai replied:
The email copy could probably use an update to include the pro-gear and pro-gear spouse. Not sure if this PR is the right place to do it.

lynzt replied:
@monfresh Donald is right... this is getting the `entitlement` to display on the email we send to the SM after they submit their move...

monfresh reviewed (approved):
LGTM. It would be good to have @transcom/truss-pamplemoose  review as well.

lynzt reviewed (commented):
LGTM... 
I have an outstanding question if we need to do anything in regards to existing moves in our system that used the old entitlement numbers or if this is only affecting moves going forward.

kimallen commented on pkg/models/shipment_summary_worksheet_test.go:
```diff
@@ -110,9 +110,11 @@ func (suite *ModelSuite) TestFetchDataShipmentSummaryWorksheet() {
-	totalEntitlement, err := models.GetEntitlement(*ssd.ServiceMember.Rank, ssd.Order.HasDependents, ssd.Order.SpouseHasProGear)
+	weightAllotment := models.GetWeightAllotment(*ssd.ServiceMember.Rank)
+	// E_9 rank, no dependents, no spouse pro-gear
+	totalWeight := weightAllotment.TotalWeightSelf + weightAllotment.ProGearWeight
```
Should this also include `weightAllotment.ProGearWeightSpouse`?

kimallen replied:
oh nm, I see the comment specifying this case

kimallen reviewed (approved):
Looks good!  Thanks for doing due-diligence and cross-team communication to make sure PPM stuff didn't get impacted.
